### PR TITLE
[ISSUE #550] Update java option setting to run in jdk9

### PR DESCRIPTION
**Pull request by husterxsp**
## What is the purpose of the change

Fix issue #550、#436.

These are some reasons for the commit.

1. JDK1.8 don't support PermSize and replace it with Metaspace.
2. JDK9 don't support -Djava.ext.dirs, and use -classpath instead.
3. JDK9 don't support PrintGCDetails, and replace it with -Xlog:gc and -Xlog:gc*
4. JDK9 don’t support UseCMSCompactAtFullCollection.
5. UseParNewGC was deprecated in Java 9 and removed in Java 10.
6. UseConcMarkSweepGC was deprecated in Java 9.
7. JDK9 Make G1 the default garbage collector
8. About `--add-exports=java.base/jdk.internal.ref=ALL-UNNAMED`
In JDK9，after starting the broker, and then type ^C to shutdown it. I got exception as follow:

```shell
^CException in thread "ShutdownHook" java.lang.IllegalStateException: java.lang.reflect.InaccessibleObjectException: Unable to make public void jdk.internal.ref.Cleaner.clean() accessible: module java.base does not "exports jdk.internal.ref" to unnamed module @55634720
    at org.apache.rocketmq.store.MappedFile$1.run(MappedFile.java:105)
    at java.base/java.security.AccessController.doPrivileged(Native Method)
    at org.apache.rocketmq.store.MappedFile.invoke(MappedFile.java:98)
    at org.apache.rocketmq.store.MappedFile.clean(MappedFile.java:94)
    at org.apache.rocketmq.store.StoreCheckpoint.shutdown(StoreCheckpoint.java:69)
    at org.apache.rocketmq.store.DefaultMessageStore.shutdown(DefaultMessageStore.java:268)
    at org.apache.rocketmq.broker.BrokerController.shutdown(BrokerController.java:748)
    at org.apache.rocketmq.broker.BrokerStartup$1.run(BrokerStartup.java:235)
    at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make public void jdk.internal.ref.Cleaner.clean() accessible: module java.base does not "exports jdk.internal.ref" to unnamed module @55634720
    at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:337)
    at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:281)
    at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:198)
    at java.base/java.lang.reflect.Method.setAccessible(Method.java:192)
    at org.apache.rocketmq.store.MappedFile$1.run(MappedFile.java:102)
    ... 8 more
```
And when I restart by run `./bin/mqbroker`, the broker just don’t start, so add `--add-exports` option to fix it.

In addition, there are still some other problem.
`/dev/shm` only exists in linux, and does not exist with mac and windows. And run on my mac will get warning: `Cannot open file /dev/shm/mq_gc_pid9236.log due to No such file or directory.`

Finally, I’ve test the commit in JDK 1.7.0_80、1.8.0_121、9.0.4、10.0.2、11.0.2 on my mac. if it’s all right, the cmd file is also need to be updated.

## Brief changelog

Update java option setting to run in jdk9.

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by zongtanghu** [comment]
@husterxsp Hi,can you resolve the failed error check in your local enviroment.

**Comment by husterxsp** [comment]
> @husterxsp Hi,can you resolve the failed error check in your local enviroment.

@zongtanghu Hi, I have just merged develop branch and all checks have passed currently. I'm not sure if it's ok now.

File: distribution/bin/runbroker.sh [hunk 1]
```diff
@@ -31,20 +31,27 @@ error_exit ()
...
 JAVA_OPT="${JAVA_OPT} -server -Xms8g -Xmx8g -Xmn4g"
 JAVA_OPT="${JAVA_OPT} -XX:+UseG1GC -XX:G1HeapRegionSize=16m -XX:G1ReservePercent=25 -XX:InitiatingHeapOccupancyPercent=30 -XX:SoftRefLRUPolicyMSPerMB=0"
-JAVA_OPT="${JAVA_OPT} -verbose:gc -Xloggc:/dev/shm/mq_gc_%p.log -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintAdaptiveSizePolicy"
-JAVA_OPT="${JAVA_OPT} -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=5 -XX:GCLogFileSize=30m"
+# JDK9 don't support PrintGCDetails, and replace it with -Xlog:gc and -Xlog:gc*
+# JDK9 don't support -Djava.ext.dirs, and use -classpath instead.
+if [[ "$JAVA_MAJOR_VERSION" -lt 9 ]] ; then
+    JAVA_OPT="${JAVA_OPT} -verbose:gc -Xloggc:/dev/shm/mq_gc_%p.log -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintAdaptiveSizePolicy"
+    JAVA_OPT="${JAVA_OPT} -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=5 -XX:GCLogFileSize=30m"
+    JAVA_OPT="${JAVA_OPT} -Djava.ext.dirs=${JAVA_HOME}/jre/lib/ext:${BASE_DIR}/lib"
```
Comment (liuruiyiyang) [inline distribution/bin/runbroker.sh]: Just for curiosity, do we need to add ```${BASE_DIR}/lib``` to ```Djava.ext.dirs``` if already added ```${BASE_DIR}/lib``` to ```CLASSPATH``` when using JDK8?

File: distribution/bin/runbroker.sh [hunk 2]
```diff
@@ -31,20 +31,27 @@ error_exit ()
...
+else
+    JAVA_OPT="${JAVA_OPT} --add-exports=java.base/jdk.internal.ref=ALL-UNNAMED"
```
Comment (liuruiyiyang) [inline distribution/bin/runbroker.sh]: Maybe we need to add some corresponding parameters of GC log for JDK9 and above here?  
e.g. ```-Xlog:gc*:file=/dev/shm/mq_gc_%p.log:timemillis:filecount=5,filesize=30720```  

And other params like PrintGCApplicationStoppedTime and PrintAdaptiveSizePolicy are also needed to be transferred in JDK9.

File: distribution/bin/runbroker.sh (hunk 1 above)
Comment (husterxsp) [reply distribution/bin/runbroker.sh]: You're right, and i've updated the commit.

File: distribution/bin/runbroker.sh (hunk 2 above)
Comment (husterxsp) [reply distribution/bin/runbroker.sh]: I‘m not very familiar with the parameters. According to [https://docs.oracle.com/javase/9/tools/java.htm#JSWOR-GUID-BE93ABDC-999C-4CB5-A88B-1994AAAC74D5](https://docs.oracle.com/javase/9/tools/java.htm#JSWOR-GUID-BE93ABDC-999C-4CB5-A88B-1994AAAC74D5)，maybe we can use `-Xlog:safepoint` and `-Xlog:ergo*=level`?

**Comment by tendant** [comment]
When can we expect this to be merged? 

This stops people from using JDK 9+.

**Comment by zongtanghu** [comment]
There is some code conflicts, in this pr, please resolve the conflicts firstyl. @husterxsp

**Pull request closed without merging**